delivery_send_to_shipper_at_operation/models/stock_picking.py

```diff
@@ -0,0 +1,90 @@
+# Copyright 2021 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+
+from lxml import etree
+
+from odoo import fields, models
+from odoo.osv import expression
+from odoo.tools.safe_eval import safe_eval
+
+from odoo.addons.base.models.ir_ui_view import (
+    transfer_modifiers_to_node,
+    transfer_node_to_modifiers,
+)
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    delivery_notification_sent = fields.Boolean(default=False)
+
+    def _send_confirmation_email(self):
+        for picking in self:
+            picking._handle_send_to_shipper_at_operation()
+            super(StockPicking, picking)._send_confirmation_email()
+
+    def _handle_send_to_shipper_at_operation(self):
+        self.ensure_one()
+        if not self.carrier_id:
+            # If the current operation has no carrier defined, but a carrier
+            # has been found in one of the chained transfers and is
+            # configured to match the current operation type: force the
+            # sending of the delivery notice to the carrier
+            related_ship = self.ship_picking_id
+            carrier = related_ship.carrier_id
+            if (
+                carrier.integration_level == "rate_and_ship"
+                and carrier.send_delivery_notice_on == "custom"
+                and self.picking_type_id
+                in carrier.send_delivery_notice_picking_type_ids
+            ):
+                self.carrier_id = carrier
+                self.send_to_shipper()
+                # Flag the current operation and the ship one.
+                # Mandatory to not execute twice 'send_to_shipper' method
+                self.delivery_notification_sent = True
+                related_ship.delivery_notification_sent = True
+
+    def send_to_shipper(self):
+        # Do not send delivery notice to the carrier if it has already been sent
+        # through a previous operation (like a pack)
+        self.ensure_one()
+        if self.delivery_notification_sent:
+            return False
+        return super().send_to_shipper()
+
+    def fields_view_get(
+        self, view_id=None, view_type="form", toolbar=False, submenu=False
+    ):
+        # Override to hide the "Send to shipper" button if the delivery
+        # notification has already been sent
+        result = super().fields_view_get(
+            view_id=view_id, view_type=view_type, toolbar=toolbar, submenu=submenu
+        )
+        if result.get("name") == "stock.picking.form":
+            result["arch"] = self._fields_view_get_adapt_send_to_shipper_attrs(
+                result["arch"]
+            )
+        return result
```

Comment:
Given how tricky this part is, it would be nice to make dovecot happy about this.

Reply:
seems a good point :)
Also, why not using the OCA module to extend existing domains?

Reply:
@simahawk A matter of preference, I'm OK to use the OCA module for custom modules, but don't like to add it as dependencies in other OCA modules, when a module touches internal things I'm always afraid to bring regressions while here it's restricted to the delivery carrier form view.